pysmt/solvers/z3.py

```diff
@@ -989,3 +998,59 @@ def sequence_interpolant(self, formulas):
 
     def _exit(self):
         pass
+
+
+class Z3NativeOptimizer(Optimizer, Z3Solver):
+
+    LOGICS = Z3Solver.LOGICS
+
+    def __init__(self, environment, logic, **options):
+        Z3Solver.__init__(self, environment=environment,
+                          logic=logic, **options)
+        self.z3 = z3.Optimize()
+
+    def _le(self, x, y):
+        otype = self.environment.stc.get_type(x)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type() or otype.is_real_type():
+            return mgr.LE(x, y)
+        elif otype.is_bv_type():
+            return mgr.BVULE(x, y)
+
+    def optimize(self, cost_function, initial_cost=None, callback=None):
+        if initial_cost is not None:
+            self.add_assertion(self._le(cost_function, initial_cost))
+        obj = self.converter.convert(cost_function)
+        h = self.z3.minimize(obj)
+
+        res = self.z3.check()
+        if res == z3.sat:
+            opt_value = self.z3.lower(h)
+            try:
+                self.converter.back(opt_value)
+                model = Z3Model(self.environment, self.z3.model())
+                return model
+            except PysmtInfinityError:
```

Comment:
Are the pySMT optimizers able to detect these two conditions? Or is this provided only by Z3?

Reply:
They are not. I added a predicate to check in the algorithm can diverge in these circumstances